packages/aws-cdk-lib/core/lib/private/runtime-info.ts

```diff
@@ -32,14 +35,68 @@ export function constructInfoFromConstruct(construct: IConstruct): ConstructInfo
     && jsiiRuntimeInfo !== null
     && typeof jsiiRuntimeInfo.fqn === 'string'
     && typeof jsiiRuntimeInfo.version === 'string') {
-    return { fqn: jsiiRuntimeInfo.fqn, version: jsiiRuntimeInfo.version };
+    return {
+      fqn: jsiiRuntimeInfo.fqn,
+      version: jsiiRuntimeInfo.version,
+      metadata: isResource(construct) ? redactTelemetryData(construct.node.metadata) : undefined,
+    };
   } else if (jsiiRuntimeInfo) {
     // There is something defined, but doesn't match our expectations. Fail fast and hard.
     throw new Error(`malformed jsii runtime info for construct: '${construct.node.path}'`);
   }
   return undefined;
 }
 
+/**
+ * Filter for Construct, Method, and Feature flag metadata. Redact values from it.
+ *
+ * @param metadata a list of metadata entries
+ */
+export function redactTelemetryData(metadata: MetadataEntry[]): Record<string, any>[] {
+  const validTypes = new Set([
+    MetadataType.CONSTRUCT,
+    MetadataType.METHOD,
+    MetadataType.FEATURE_FLAG,
+  ]);
+
+  return metadata
+    .filter((entry) => validTypes.has(entry.type as MetadataType))
+    .map((entry) => ({
+      type: entry.type,
+      data: redactTelemetryDataHelper(entry.data),
+    }));
+}
+
+/**
+ * Redact values from dictionary values other than Boolean and ENUM-type values.
+ * @TODO complete the ENUM-type values redaction in a follow-up change.
+ */
+function redactTelemetryDataHelper(data: any): any {
+  if (typeof data === 'boolean') {
+    return data; // Return booleans as-is
+  }
+
+  if (Array.isArray(data)) {
+    // Handle arrays by recursively redacting each element
+    return data.map((item) => redactTelemetryDataHelper(item));
+  }
+
+  if (data && typeof data === 'object') {
+    // Handle objects by iterating over their key-value pairs
+    if (isResource(data)) {
+      return '*';
```

Comment:
can you explain why you decided that we should overwrite the value if it is referring to a resource. Also, do you know if the `isResource` function check both L1 and L2 resources.

Reply:
The design mentions that we will not collect L1 usage. `isResource` will filter out L1 construct usages. The reason that we override the value if it's a L2 resource is because this resource is already included in the `metadata` and we don't want to include duplicate resource logging.